transform_constantfold.ml

```diff
@@ -69,53 +62,91 @@ let const_prop ({formals; instrs} : analysis_input) : instructions option =
-    let open Analysis in
-    match worklist with
-    | [] -> acc
-    | pc :: rest ->
-      begin match[@warning "-4"] instrs.(pc) with
-      | Drop y when x = y ->
-        (* `x` is now out of scope, but continue searching the `rest` of the
-           worklist. *)
-        find_targets instrs x rest acc
-      | instr ->
-        if PcSet.mem pc acc then
-          (* Already seen this pc, but continue searching the `rest` of the
-             worklist. *)
-          find_targets instrs x rest acc
-        else
-          (* Add the current `pc` to the accumulator and update the worklist
-             with our successors. *)
-          let succs = successors_at instrs pc in
-          find_targets instrs x (succs @ rest) (PcSet.add pc acc)
-      end
-  in
+  let module VarMap = Map.Make(Variable) in
+
+  let module Approx = struct
+    type t = Unknown | Value of value
+    let equal a1 a2 = match a1, a2 with
+      | Unknown, Unknown -> true
+      | Unknown, Value _ | Value _, Unknown -> false
+      | Value v1, Value v2 -> Eval.value_eq v1 v2
+  end in
 
-  (* Perform constant propagation. *)
-  let work instrs =
+  (* for each instruction, what is the set of variables that are constant? *)
+  let constants =
     let open Analysis in
-    (* Find all constant propagation candidates. *)
-    let candidates = find_candidates instrs 0 [] in
-    if candidates = [] then None else
-    (* Perform all propagations for a single candidate. *)
-    let propagate instrs (pc, x, l) =
-      let succs = successors_at instrs pc in
-      let targets = find_targets instrs x succs PcSet.empty in
-      let instrs = Array.copy instrs in
-      let convert_at pc = instrs.(pc) <- convert x (Constant(l)) instrs.(pc) in
-      PcSet.iter convert_at targets;
-      instrs
+    let open Approx in
+    let merge pc cur incom =
+      let merge_approx x mv1 mv2 = match mv1, mv2 with
+        | None, _ | _, None ->
+          failwith "scope error?"
+        | Some Unknown, _ | _, Some Unknown -> Some Unknown
+        | Some (Value v1), Some (Value v2) ->
+          if Eval.value_eq v1 v2 then Some (Value v1) else Some Unknown
+      in
+      if VarMap.equal Approx.equal cur incom then None
+      else Some (VarMap.merge merge_approx cur incom)
     in
-    Some (List.fold_left propagate instrs candidates)
+    let update pc cur =
+      match[@warning "-4"] instrs.(pc) with
+      | Decl_const (x, Simple(Constant l))
+      | Decl_mut (x, Some (Simple (Constant l)))
+        ->
+        VarMap.add x (Value l) cur
+      | Decl_array (x, _) ->
+        (* this case could be improved with approximation for arrays so
+           that, eg., length(x) could be constant-folded *)
+        VarMap.add x Unknown cur
+      | Decl_const (x, _) | Decl_mut (x, _) ->
+        VarMap.add x Unknown cur
+      | Call (x, _, _) as call ->
+        cur
+        |> ModedVarSet.fold (fun (_, x) -> VarMap.remove x) (changed_vars call)
```

Comment:
Is there a reason the changed variables are removed, rather than setting them to `Unknown`? In other words, what is the difference between `Unknown` and not being in the map? (More comments would be helpful for answering the second question.)